File: src/fixpm/corrector.py
"""Turn analyzer issues into concrete, ranked replacement commands."""

from __future__ import annotations

import shlex

from .detector import analyze, tokenize
from .distance import similarity
from .packages import NpmRegistry, base_name, format_downloads
from .rules.base import Correction, Issue, IssueKind, ManagerSpec

_client: NpmRegistry | None = None

# Placeholder shown when a fix needs manual input (never auto-executed).
_PLACEHOLDERS = {
    "run": "<script>",
    "exec": "<package> [args...]",
    "dlx": "<package> [args...]",
    "add": "<package>",
    "install": "<package>",
    "remove": "<package>",
    "uninstall": "<package>",
    "why": "<package>",
    "info": "<package>",
    "create": "<initializer>",
    "config": "<key> <value>",
    "cache": "<action>",
    "pkg": "<action>",
}


def _default_client() -> NpmRegistry:
    global _client
    if _client is None:
        _client = NpmRegistry()
    return _client


def get_corrections(
    text: str,
    *,
    spec: ManagerSpec | None = None,
    client: NpmRegistry | None = None,
) -> list[Correction]:
    """Top fix suggestions (max 3), best first."""
    tokens, spec, issues = analyze(text, spec)
    if spec is None or not issues:
        return []
    client = client or _default_client()
    corrections: list[Correction] = []
    for issue in issues:
        corrections.extend(_fix_issue(tokens, spec, issue, client))
    best: dict[str, Correction] = {}
    for c in corrections:
        if c.command not in best or c.score > best[c.command].score:
            best[c.command] = c
    ranked = sorted(best.values(), key=lambda c: (-c.score, c.command))
    return ranked[:3]


def _ratio_score(ratio: float) -> float:
    return round(0.5 + 0.45 * ratio, 3)


def _replace(tokens: list[str], index: int, new_token: str) -> str:
    patched = list(tokens)
    patched[index] = new_token
    return shlex.join(patched)


def _fix_issue(tokens: list[str], spec: ManagerSpec, issue: Issue,
               client: NpmRegistry) -> list[Correction]:
    kind = issue.kind

    if kind is IssueKind.SUBCOMMAND_TYPO:
        out = []
        hinted = spec.typo_hints.get(issue.token.lower())
        for cand in issue.candidates[:3]:
            score = (0.95 if cand == hinted
                     else _ratio_score(similarity(issue.token.lower(), cand)))
            out.append(Correction(_replace(tokens, issue.index, cand),
                                  kind, score, issue.message))
        return out

    if kind is IssueKind.FLAG_TYPO:
        return [
            Correction(_replace(tokens, issue.index, cand), kind,
                       _ratio_score(similarity(issue.token, cand)),
                       issue.message)
            for cand in issue.candidates[:2]
        ]

    if kind is IssueKind.MISSING_DASHES:
        return [Correction(_replace(tokens, issue.index, issue.candidates[0]),
                           kind, 0.9, issue.message)]

    if kind is IssueKind.ARG_REQUIRED:
        placeholder = _PLACEHOLDERS.get(issue.token, "<arguments>")
        return [Correction(f"{shlex.join(tokens)} {placeholder}", kind, 0.6,
                           issue.message, False)]

    if kind is IssueKind.PACKAGE_TYPO:
        name = base_name(issue.token)
        out = []
        for s in client.suggest(name)[:3]:
            if s.name.lower() == name.lower():
                continue
            new_token = issue.token.replace(name, s.name, 1)
            detail = (f"package '{name}' -> '{s.name}' "
                      f"({format_downloads(s.weekly_downloads)}/week)")
            out.append(Correction(_replace(tokens, issue.index, new_token),
                                  kind, s.score, detail))
        return out

    return []


__all__ = ["get_corrections", "tokenize"]

File: go/corrector.go
// Turn analyzer issues into concrete, ranked replacement commands — port of
// src/fixpm/corrector.py.
package main

import (
	"sort"
	"strings"
	"time"
)

// placeholders mirrors corrector._PLACEHOLDERS: the visible placeholder shown
// when a fix needs manual input (such fixes are never auto-executed).
var placeholders = map[string]string{
	"run":       "<script>",
	"exec":      "<package> [args...]",
	"dlx":       "<package> [args...]",
	"add":       "<package>",
	"install":   "<package>",
	"remove":    "<package>",
	"uninstall": "<package>",
	"why":       "<package>",
	"info":      "<package>",
	"create":    "<initializer>",
	"config":    "<key> <value>",
	"cache":     "<action>",
	"pkg":       "<action>",
}

// Matches the Python default timeout of 2.5s per request.
var defaultRegistry = NewNpmRegistry(2500 * time.Millisecond)

// ratioScore mirrors corrector._ratio_score.
func ratioScore(ratio float64) float64 {
	return PyRound(0.5+0.45*ratio, 3)
}

// replaceToken mirrors corrector._replace: patch one token and re-join.
func replaceToken(tokens []string, index int, newToken string) string {
	if index < 0 || index >= len(tokens) {
		return ShlexJoin(tokens)
	}
	patched := make([]string, len(tokens))
	copy(patched, tokens)
	patched[index] = newToken
	return ShlexJoin(patched)
}

// GetCorrections mirrors corrector.get_corrections: at most 3 fixes, best first.
//
// A nil registry falls back to the package-level default, matching the Python
// `client = client or _default_client()`.
func GetCorrections(text string, spec *ManagerSpec, client *NpmRegistry) []Correction {
	tokens, found, issues := Analyze(text, spec)
	if found == nil || len(issues) == 0 {
		return nil
	}
	if client == nil {
		client = defaultRegistry
	}

	all := []Correction{}
	for _, issue := range issues {
		all = append(all, fixIssue(tokens, found, issue, client)...)
	}

	// Dedupe by command keeping the highest score. Python relies on dict
	// insertion order, so the first command seen keeps its rank position.
	best := map[string]Correction{}
	order := []string{}
	for _, c := range all {
		prev, seen := best[c.Command]
		if !seen {
			best[c.Command] = c
			order = append(order, c.Command)
			continue
		}
		if c.Score > prev.Score {
			best[c.Command] = c
		}
	}

	ranked := make([]Correction, 0, len(order))
	for _, cmd := range order {
		ranked = append(ranked, best[cmd])
	}
	sort.SliceStable(ranked, func(i, j int) bool {
		if ranked[i].Score != ranked[j].Score {
			return ranked[i].Score > ranked[j].Score
		}
		return ranked[i].Command < ranked[j].Command
	})
	if len(ranked) > 3 {
		ranked = ranked[:3]
	}
	return ranked
}

// fixIssue mirrors corrector._fix_issue.
func fixIssue(tokens []string, spec *ManagerSpec, issue Issue,
	client *NpmRegistry) []Correction {

	switch issue.Kind {
	case KindSubcommandTypo:
		out := []Correction{}
		hinted, hasHint := spec.TypoHints[strings.ToLower(issue.Token)]
		for _, cand := range head(issue.Candidates, 3) {
			score := ratioScore(Similarity(strings.ToLower(issue.Token), cand))
			if hasHint && cand == hinted {
				score = 0.95
			}
			out = append(out, Correction{
				Command:    replaceToken(tokens, issue.Index, cand),
				Kind:       issue.Kind,
				Score:      score,
				Message:    issue.Message,
				Executable: true,
			})
		}
		return out

	case KindFlagTypo:
		out := []Correction{}
		for _, cand := range head(issue.Candidates, 2) {
			out = append(out, Correction{
				Command:    replaceToken(tokens, issue.Index, cand),
				Kind:       issue.Kind,
				Score:      ratioScore(Similarity(issue.Token, cand)),
				Message:    issue.Message,
				Executable: true,
			})
		}
		return out

	case KindMissingDashes:
		if len(issue.Candidates) == 0 {
			return nil
		}
		return []Correction{{
			Command:    replaceToken(tokens, issue.Index, issue.Candidates[0]),
			Kind:       issue.Kind,
			Score:      0.9,
			Message:    issue.Message,
			Executable: true,
		}}

	case KindArgRequired:
		placeholder, ok := placeholders[issue.Token]
		if !ok {
			placeholder = "<arguments>"
		}
		return []Correction{{
			Command:    ShlexJoin(tokens) + " " + placeholder,
			Kind:       issue.Kind,
			Score:      0.6,
			Message:    issue.Message,
			Executable: false, // needs manual input: never auto-executed
		}}

	case KindPackageTypo:
		name := baseName(issue.Token)
		out := []Correction{}
		for _, s := range client.Suggest(name, 3) {
			if strings.EqualFold(s.Name, name) {
				continue
			}
			newToken := strings.Replace(issue.Token, name, s.Name, 1)
			out = append(out, Correction{
				Command: replaceToken(tokens, issue.Index, newToken),
				Kind:    issue.Kind,
				Score:   s.Score,
				Message: "package '" + name + "' -> '" + s.Name + "' (" +
					formatDownloads(s.WeeklyDownloads) + "/week)",
				Executable: true,
			})
		}
		return out
	}
	return nil
}

// head mirrors Python's `seq[:n]` for a slice.
func head(items []string, n int) []string {
	if len(items) > n {
		return items[:n]
	}
	return items
}
